Centre the main window (Windows)

logiclrd opened the pull request:
When the Windows app is starting, it places the window with a specific location and size. That location is currently hardcoded to be the pixel offset (10, 10). Most parts of the system, including all subsidiary windows, automatically centre themselves. This PR updates the top-level window to be centered as well.

This PR is built on top of #12933. If that PR gets merged first, I will rebase this one as needed. If this PR gets merged instead of that PR, the changes from that PR will be included via this PR. As of creation, the diff on this PR shows the combined effect of both PRs, but this PR properly only includes this commit:

https://github.com/rustdesk/rustdesk/commit/04cc078afe64387a94253a213332f57e56e5bf7d

fufesou commented:
I think https://github.com/rustdesk/rustdesk/pull/12933 is necessary, as it does correct the previously incorrect calculation.

But why isn't Dart used in this PR? This is a cross-platform issue, so wouldn't Dart be more appropriate for this modification?
We also use `WindowManager.center()` for the other windows, "Remote windows", "Filter transfer window" and "Installation window"...
It's probably better to modify it within the window manager.

In fact, if it's centered, then an `800 x 600` window should display good on a typical monitor, although the taskbar offset is not excluded.

logiclrd commented:
Rebased. There is now a utility function that encapsulates the logic for centering a window on the monitor that contains it/is nearest.

logiclrd committed:
Reworked the Windows runner/main.cpp to centre the window. Added function CentreInWorkArea to win32_desktop.cpp/.h.

logiclrd commented:
I have rebased this branch, which was previously based on #12933 and thus duplicated its changes, onto `master`, into which those base changes are now merged, leaving only the window-centering commit. 🙂

logiclrd commented:
Huh. I'm not entirely certain, but I did some more testing and it looks like it might be down to a delay between the window being moved and the new position being saved. I just ran a test where I ran RustDesk, moved the window, and then immediately closed the window with Alt-F4 to minimize the delay. Back to (10, 10). Did it again. Back to (10, 10). Then I ran it, moved the window, waited 3 seconds, and closed it. Now that position is remembered.

logiclrd committed:
Enhappify static analysis. :-)

logiclrd commented:
> #12974 

#12987 instead.

rustdesk commented:
https://github.com/rustdesk/rustdesk/pull/13003
